### PR TITLE
feat(cli): support SSO

### PR DESCRIPTION
**Pull request by comcalvi**
Adds support for SSO.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts [hunk 1]
```diff
@@ -53,6 +54,11 @@ export class AwsCliCompatible {
       () => new AWS.EnvironmentCredentials('AMAZON'),
     ];
 
+    if (process.env.AWS_PROFILE) {
+      await forceSdkToReadConfigIfPresent();
+      sources.push(() => new AWS.SsoCredentials({ profile: implicitProfile }));
+    }
+
     if (await fs.pathExists(credentialsFileName())) {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts:62]: Can you try it without this `if`, and putting the `SsoCredentials` inside the `sources` array below?

File: packages/aws-cdk/THIRD_PARTY_LICENSES [hunk 2]
```diff
@@ -3241,6 +3241,16 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 
+----------------
```
Comment (rix0rrr) [inline packages/aws-cdk/THIRD_PARTY_LICENSES:3244]: I want Eli to have a look at this.

**Review by rix0rrr — APPROVED** [review]

File: packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts (hunk 1 above)
Comment (comcalvi) [reply packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts]: @rix0rrr yep, this change works, with both an existing (but empty) credentials file and with no credentials file. Removing the config file breaks it, but that's expected.

File: packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts [hunk 3]
```diff
@@ -53,13 +54,12 @@ export class AwsCliCompatible {
...
-      // Force reading the `config` file if it exists by setting the appropriate
-      // environment variable.
-      await forceSdkToReadConfigIfPresent();
-      sources.push(() => profileCredentials(implicitProfile));
-      sources.push(() => new AWS.ProcessCredentials({ profile: implicitProfile }));
-    }
+    // Force reading the `config` file if it exists by setting the appropriate
+    // environment variable.
+    await forceSdkToReadConfigIfPresent();
+    sources.push(() => profileCredentials(implicitProfile));
```
Comment (eladb) [inline packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts:60]: Seems like the list of sources is duplicated in two places in the code. Maybe find a way to reuse?

File: packages/aws-cdk/THIRD_PARTY_LICENSES (hunk 2 above)
Comment (iliapolo) [reply packages/aws-cdk/THIRD_PARTY_LICENSES]: I'm not sure why, but looks like the `aws-sdk` upgrade in the CLI package broke the hoisting.

Currently on master we have:

```console
<repoRoot>/node_modules/aws-sdk (version 2.1094.0)
```

When I upgraded `aws-sdk@^2.979.0` to `aws-sdk@^2.1093.0` and ran `yarn install`, I got the following directory structure:

```console
<repoRoot>/node_modules/aws-sdk (version 2.1094.0)
<repoRoot>/packages/aws-cdk/node_modules/aws-sdk (version 2.1096.0)
```

I expected version 2.1096.0 to be hoisted and replace 2.1094.0 - but thats not what happens. 

So in practice the CLI uses two different versions of `aws-sdk`. One coming from its own direct dependency, which isn't hoisted anymore (2.1096.0), and one coming transitively from `cdk-assets`, which is hoisted (2.1094.0).

We need to figure out why yarn isn't hoisting it. @rix0rrr Any ideas?

Comment (rix0rrr) [reply packages/aws-cdk/THIRD_PARTY_LICENSES]: No, feels like it should be able to. Wonder if there's a constraint somewhere that's preventing the hoist?

File: packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts [hunk 4]
```diff
@@ -83,6 +74,14 @@ export class AwsCliCompatible {
         tokenCodeFn,
       });
     }
+
+    function iniFileCredentialFactories(theProfile: string) {
+      return [
+        () => profileCredentials(theProfile),
```
Comment (pgrzesik) [inline packages/aws-cdk/lib/api/aws-auth/awscli-compatible.ts:80]: Hello, is there a reason why credentials from the profile would take precedence here over the SSO? It's different than e.g. behavior of JS SDK v3 - https://github.com/aws/aws-sdk-js-v3/blob/31e72ac6063fdb3393c2d5042b8964238ce1b23a/packages/credential-provider-node/src/defaultProvider.ts#L52-L54

Or am I missing something here?

**Pull request merged**